compio-io/src/write/mod.rs

```diff
@@ -181,15 +159,45 @@ macro_rules! impl_write_at {
 impl_write_at!(@ptr &mut A, Box<A>);
 impl_write_at!(@slice [u8], const LEN => [u8; LEN]);
 
-impl AsyncWriteAt for Vec<u8> {
+impl<#[cfg(feature = "allocator_api")] A: Allocator> AsyncWriteAt for vec_alloc!(u8, A) {
     async fn write_at<T: IoBuf>(&mut self, buf: T, pos: u64) -> BufResult<usize, T> {
         let pos = pos as usize;
         let slice = buf.as_slice();
-        let n = slice.len().min(self.len() - pos);
-        self[pos..pos + n].copy_from_slice(&slice[..n]);
-        if n < slice.len() {
-            self.extend_from_slice(&slice[n..]);
+        if pos <= self.len() {
+            let n = slice.len().min(self.len() - pos);
+            self[pos..pos + n].copy_from_slice(&slice[..n]);
+            if n < slice.len() {
```

Comment:
Determine whether resize is required beforehand so that there's only one realloc. It's a three branch situation:
- pos < end and available space >= buf_len
- pos < end but available space < buf_len
- pos >= end